atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/OracleTableNameMapper.java

```diff
@@ -0,0 +1,99 @@
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.DbKvs;
+import com.palantir.nexus.db.sql.AgnosticResultSet;
+
+public class OracleTableNameMapper {
+    public static final int ORACLE_MAX_TABLE_NAME_LENGTH = 30;
+    public static final int SUFFIX_NUMBER_LENGTH = 6;
+    public static final int MAX_NAMESPACE_LENGTH = 2;
+    private static final int PREFIXED_TABLE_NAME_LENGTH = ORACLE_MAX_TABLE_NAME_LENGTH - SUFFIX_NUMBER_LENGTH;
+
+    private final ConnectionSupplier conns;
+
+    public OracleTableNameMapper(ConnectionSupplier conns) {
+        this.conns = conns;
+    }
+
+    public String getShortPrefixedTableName(String tablePrefix, TableReference tableRef) {
+        Preconditions.checkState(tablePrefix.length() <= 7, "The tablePrefix can be at most 7 characters long");
+
+        String fullTableName = tablePrefix + DbKvs.internalTableName(tableRef);
+        if (fullTableName.length() <= PREFIXED_TABLE_NAME_LENGTH) {
+            return fullTableName;
+        }
+
+        TableReference shortenedNamespaceTableRef = truncateNamespace(tableRef);
+
+        String prefixedTableName = tablePrefix + DbKvs.internalTableName(shortenedNamespaceTableRef);
+        String truncatedTableName = truncate(prefixedTableName, PREFIXED_TABLE_NAME_LENGTH);
+        return truncatedTableName + getTableNumberSuffix(fullTableName, truncatedTableName);
+    }
+
+    private TableReference truncateNamespace(TableReference tableRef) {
+        if (tableRef.getNamespace().isEmptyNamespace()) {
+            return tableRef;
+        }
+        String namespace = tableRef.getNamespace().getName();
+        namespace = truncate(namespace, MAX_NAMESPACE_LENGTH);
+        return TableReference.create(Namespace.create(namespace), tableRef.getTablename());
+    }
+
+    private String truncate(String name, int length) {
+        return name.substring(0, Math.min(name.length(), length));
+    }
+
+    private String getTableNumberSuffix(String fullTableName, String truncatedTableName) {
+        int tableSuffixNumber = getNextTableNumber(truncatedTableName);
+        if (tableSuffixNumber >= 100_000) {
+            throw new IllegalArgumentException(
+                    "Cannot create any more tables with name starting with " + truncatedTableName
+                    + ". 100,000 tables might have already been created. Please rename the table." + fullTableName);
+        }
+        return "_" + String.format("%05d", tableSuffixNumber);
+    }
+
+    private int getNextTableNumber(String truncatedTableName) {
+        AgnosticResultSet results = conns.get().selectResultSetUnregisteredQuery(
+                "SELECT short_table_name "
+                + "FROM " + AtlasDbConstants.ORACLE_NAME_MAPPING_TABLE
+                + " WHERE LOWER(short_table_name) LIKE LOWER('" + truncatedTableName + "!______%') ESCAPE '!'"
+                + " ORDER BY short_table_name DESC");
+        if (results.size() == 0) {
```

Comment:
Why do we special case this? `getTableNumberWhenPrefixedTableExist` just doesn't do a `for` loop